src/frame/content.rs

```diff
@@ -193,6 +251,56 @@ pub struct SynchronisedLyrics {
     pub content: Vec<(u32, String)>,
 }
 
+const MILLISECONDS_PER_HOUR: u32 = 3600000;
+const MILLISECONDS_PER_MINUTE: u32 = 60000;
+const MILLISECONDS_PER_SECOND: u32 = 1000;
+
+impl SynchronisedLyrics {
+    /// Write the lyrics to the provided `writer` as a plain text table.
+    ///
+    /// A typical table might look like:
+    ///
+    /// ```text
+    /// Timecode        Lyrics
+    /// 00:00:12.123    Song line one
+    /// 00:00:22.456    Song line two
+    /// …
+    ///```
+    ///
+    /// # Errors
+    ///
+    /// This function will return any I/O error reported while formatting.
+    pub fn fmt_table(&self, writer: &mut impl io::Write) -> io::Result<()> {
```

Comment:
The `&mut` is not needed, `io::Write` is implemented for `&mut io::Write` :)

Reply:
Hello,

I seem to need a mutable reference. Unless I take ownership of `writer`…

```
error[E0596]: cannot borrow `*writer` as mutable, as it is behind a `&` reference
   --> src/frame/content.rs:276:24
    |
273 |     pub fn fmt_table(&self, writer: &impl io::Write) -> io::Result<()> {
    |                                     --------------- help: consider changing this to be a mutable reference: `&mut impl io::Write`
...
276 |                 write!(writer, "Frame\t{}\n", self.content_type)?;
    |                        ^^^^^^ `writer` is a `&` reference, so the data it refers to cannot be borrowed as mutable
```

Can you help me understand what I am doing wrong?

Thanks for your help.

Dan

Reply:
Remove the `&`, you do not need any kind of reference for `writer`.

`io::Write` is implemented for any mutable reference to an `io::Write`. So callers of `fmt_table` can pass a `&mut io::Write` without this being required the type signature.

If any function in the body of `fmt_table` moves `writer`, you can just use `&mut writer` inline, but iirc the `write!` macro does not move `writer` so no need to change anything there.

Reply:
Sorry to say, I still haven't got this to work :disappointed: 

When I remove the `&` and try and use the `write!` macro without an inline `&mut writer` reference like this (elisions indicated by `…`):
```rust
 pub fn fmt_table(&self, writer: impl io::Write) -> io::Result<()> {
        match self.timestamp_format {
            TimestampFormat::MPEG => {
                write!(writer, "Frame\t{}\n", self.content_type)?;
…
 Ok(())
    }
```
I get the following error:
```rust
error[E0596]: cannot borrow `writer` as mutable, as it is not declared as mutable
   --> src/frame/content.rs:276:24
    |
273 |     pub fn fmt_table(&self, writer: impl io::Write) -> io::Result<()> {
    |                             ------ help: consider changing this to be mutable: `mut writer`
...
276 |                 write!(writer, "Frame\t{}\n", self.content_type)?;
    |                        ^^^^^^ cannot borrow as mutable
```

Adding an inline `&mut writer` for every call of the `write!` macro, like this:
```rust
pub fn fmt_table(&self, writer: impl io::Write) -> io::Result<()> {
        match self.timestamp_format {
            TimestampFormat::MPEG => {
                write!(&mut writer, "Frame\t{}\n", self.content_type)?;
…
 Ok(())
    }
```
yields the following error:
```rust
error[E0596]: cannot borrow `writer` as mutable, as it is not declared as mutable
   --> src/frame/content.rs:276:24
    |
273 |     pub fn fmt_table(&self, writer: impl io::Write) -> io::Result<()> {
    |                             ------ help: consider changing this to be mutable: `mut writer`
...
276 |                 write!(&mut writer, "Frame\t{}\n", self.content_type)?;
    |                        ^^^^^^^^^^^ cannot borrow as mutable
```

I am using the latest stable version of rust.
```shell
> rustc --version
rustc 1.43.0 (4fb7144ed 2020-04-20)
```

:confused:

Reply:
Whoops sorry, think I might have it now!

In 7cf300c I have changed the signature to:
```rust
pub fn fmt_table(&self, mut writer: impl io::Write) -> io::Result<()>
```
and that seems to work. :smiley: